### ISSUE TITLE
Compiler flags that trap uninitialized variables

### ISSUE THREAD

**Issue by rgknox** [ESMCI/cime#3205]
With the FATES model we have benefited from running a few tests with compiler flags that catch uninitialized variables.  On intel compilers this is "-ftrapuv".  I believe that it auto-initializes all variables to some type of nan'ish-equivalent (?), so if it is used without being explicitly set, it crash upon attempted use.

I just wanted to throw this thought out to the ether to see if others are doing this as well, or perhaps queue others if they have had similar intents.

@glemieux

**Comment by ekluzek** [comment]
On cheyenne when DEBUG=TRUE "-check uninit" is being set, which is just slightly different. I think we used to use "-ftrapuv", but maybe there were problems with it?

The intel man page defines them each as follows:

>               check uninit      Enables run-time checking for uninitialized variables. If a variable is read before it is written, a run-time error routine will be
>                                 called. Only local scalar variables of intrinsic type INTEGER, REAL, COMPLEX, and LOGICAL without the SAVE attribute are checked.

and

>        -ftrapuv
> 
>               Initializes stack local variables to an
>                      unusual value to aid error detection.
>

**Comment by glemieux** [comment]
I found this Intel article discussing the differences between `-ftrapuv` and `-check uninit`and why y'all might not be using the former: https://software.intel.com/en-us/articles/dont-optimize-when-using-ftrapuv-for-uninitialized-variable-detection.  

@ekluzek, this came out of some troubleshooting: I've been trying to figure out a RUN failure that crops up on Lawrencium (https://github.com/NGEET/fates/issues/548), but not Cheyenne, that only seems to come up when `DEBUG=TRUE`.  The offending bit of code seems to be a call to an uninitialized global variable, `hlm_model_day`.  @rgknox pointed out to me that he had added the `-ftrapuv` option to the debug configuration for Lawrencium in config_compilers.xml, which was likely catching it.  I tested this out by removing the `-ftrapuv` option and was able to successfully run debug mode tests on Lawrencium.  I'm pretty sure that `hlm_model_day` is being called when uninitialized, so I'm not sure why `-check uninit` isn't catching it.  Maybe I'm misunderstanding the flow of the calls here...
